kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1522 @@
+    * @brief This method assembles the global relation matrix (T matrix used to impose the MPC)
+    * @param rT The global relation matrix
+    * @param rTransformationMatrix The local transformation contribution
+    * @param rSlaveEquationId The equation id of the slave dofs
+    * @param rMasterEquationId The equation id of the master dofs
+    * @param rLockArray The lock of the dof
+    */
+    void AssembleRelationMatrix(
+        TSystemMatrixType& rT,
+        const LocalSystemMatrixType& rTransformationMatrix,
+        const EquationIdVectorType& rSlaveEquationId,
+        const EquationIdVectorType& rMasterEquationId
+#ifdef USE_LOCKS_IN_ASSEMBLY
+        ,std::vector< omp_lock_t >& rLockArray
+#endif
+        )
+    {
+        SizeType local_size_1 = rTransformationMatrix.size1();
+
+        for (IndexType i_local = 0; i_local < local_size_1; ++i_local) {
+            IndexType i_global = rSlaveEquationId[i_local];
+
+            if (i_global < BaseType::mEquationSystemSize) {
+            #ifdef USE_LOCKS_IN_ASSEMBLY
+                omp_set_lock(&rLockArray[i_global]);
+            #endif
+
+                BaseType::AssembleRowContributionFreeDofs(rT, rTransformationMatrix, i_global, i_local, rMasterEquationId);
+
+            #ifdef USE_LOCKS_IN_ASSEMBLY
+                omp_unset_lock(&rLockArray[i_global]);
+            #endif
+            }
+            //note that computation of reactions is not performed here!
+        }
+    }
+
+    /**
+     * @brief This method construcs the relationship between the DoF
+     * @param pScheme The integration scheme
+     * @param rA The LHS of the system
+     * @param rModelPart The model part which defines the problem
+     */
+    void ConstructMatrixStructure(
+        typename TSchemeType::Pointer pScheme,
+        TSystemMatrixType& rA,
+        ModelPart& rModelPart
+        ) override
+    {
+        if(rModelPart.MasterSlaveConstraints().size() > 0)
+            ConstructMatrixStructureWithConstraints(pScheme, rA, rModelPart);
+        else
+            BaseType::ConstructMatrixStructure(pScheme, rA, rModelPart);
+    }
+
+    /**
+     * @brief The same methods as the base class but with constraints
+     * @param pScheme The pointer to the integration scheme
+     * @param rModelPart The model part to compute
+     * @param rA The LHS matrix of the system of equations
+     * @param rDx The vector of unkowns
+     * @param rb The RHS vector of the system of equations
+     */
+    void BuildAndSolveWithConstraints(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb
+        )
+    {
+        KRATOS_TRY
+
+        Timer::Start("Build");
+
+        // We do the build (after that we resize the solution vector to avoid problems)
+        Build(pScheme, rModelPart, rA, rb);
+        rDx.resize(mDoFToSolveSystemSize, false);
+
+        Timer::Stop("Build");
```

Comment:
shouldn't this stop be before line 448 ?

Reply:
Not in my opinion, I am counting building times everything, not just the assembling

Reply:
Its upto you ... but have a look here ..
https://github.com/KratosMultiphysics/Kratos/blob/58d438677a7cf5a2909d79b76bcbd70290e0cc2e/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L407

resizing is not building.

Reply:
Well in this specific case I think is more open, what do you thing @RiccardoRossi